feat(otlp): Derive a `sentry.description` attribute for V2 spans

gggritso opened the pull request:
Very similar to https://github.com/getsentry/relay/pull/4796, but for the `span.description` attribute. Supports [Span Field Transition](https://linear.app/getsentry/document/span-field-transition-e0e4b8e59f11).

V2 spans probably will still have a description, but OTLP spans will not. To make the transition easier, and to create a better experience in the span waterfall, this PR derives a description for HTTP and database spans based on the logic that exists in the [JavaScript SDK](https://github.com/getsentry/sentry-javascript/blob/master/packages/opentelemetry/src/utils/parseSpanDescription.ts#L41-L94). It's _slightly_ different from the SDK logic, but not in any important way, from what I can tell.

1. If the payload includes a description, obeys the description as sent, even if it's an empty string, and sets it as the `sentry.description` attribute
2. If the payload does not include a `description` field, generate a description and put it in the `sentry.description` attribute
3. If we cannot determine a good description, fall back to using the `name` field, since it's similar in purpose

I also removed the little bit of current `description` processing, since it all lives near the `op` processing now.

gggritso committed:
Claude attempt

gggritso committed:
Remove unnecessary argument

gggritso committed:
Improve HTTP description fallback

If an HTTP method is available, use it

gggritso committed:
Remove unnecessary return parameter

gggritso committed:
Prevent passing an entire span for no reason

gggritso committed:
Tidy up URL formatting

gggritso committed:
Don't pass pointless variable

gggritso committed:
Clean up default return

gggritso committed:
Add scheme delimiter

gggritso committed:
Expand tests

gggritso committed:
Update snapshots

gggritso committed:
A bit more documentation

gggritso committed:
Remove existing processing

gggritso committed:
Fix comment indentation

gggritso committed:
Update snapshots

gggritso committed:
Add changelog entry

gggritso committed:
Always fall back to `span.name`

gggritso committed:
Pull from `db.query.text`

gggritso commented on relay-spans/src/v2_to_v1.rs:
```diff
@@ -270,6 +277,178 @@ fn derive_op_for_v2_span(span: &SpanV2) -> String {
+            .get("sentry.op")
+            .and_then(|attr| attr.value())
+            .and_then(|attr_val| attr_val.value.value.value())
+            .and_then(|v| v.as_str())
```
This pattern is starting to show up a _lot_. Maybe worth adding a method like `dig` to `Attributes` to pull this data out?

loewenheim replied:
See https://github.com/getsentry/relay/pull/4838.

gggritso committed:
Update spanshots

loewenheim commented on relay-spans/src/otel_to_sentry.rs:
```diff
@@ -11,6 +11,9 @@ use relay_protocol::Error;
 ///   if the OTEL span's `name` is empty.
 /// * The Sentry span's `op` field will be inferred based on the OTEL span's `sentry.op` attribute,
 ///   or other available attributes if `sentry.op` is not provided.
+/// * The Sentry span's `description` field will be inferred based on the OTEL span's
+///   `sentry.description` attribute, or other available attributes if `sentry.description` is not
+///   provided.
```
The line immediately below already refers to the `description`.

loewenheim commented on relay-spans/src/otel_to_sentry.rs:
```diff
@@ -339,7 +351,7 @@ mod tests {
             "spanId": "e342abb1214ca181",
             "parentSpanId": "0c7a7dea069bf5a6",
             "name": "http client request",
-            "kind": 3,
+            "kind": 2,
```
Is there a reason this is changing?

loewenheim commented on relay-spans/src/v2_to_v1.rs:
```diff
@@ -270,6 +277,178 @@ fn derive_op_for_v2_span(span: &SpanV2) -> String {
+        return description;
+    };
+
+    if attributes.contains_key("http.request.method") || attributes.contains_key("http.method") {
```
This check is redundant, `derive_http_description` returns `None` anyway if these are both unset.

loewenheim commented on relay-spans/src/v2_to_v1.rs:
```diff
@@ -270,6 +277,178 @@ fn derive_op_for_v2_span(span: &SpanV2) -> String {
+            .and_then(|attr_val| attr_val.value.value.value())
+            .and_then(|v| v.as_str())
+            .map(|op| op.starts_with("cache."))
+            .unwrap_or(false);
```
```suggestion
            .is_some_and(|op| op.starts_with("cache."));
```

loewenheim commented on relay-spans/src/v2_to_v1.rs:
```diff
@@ -270,6 +277,178 @@ fn derive_op_for_v2_span(span: &SpanV2) -> String {
+    }
+
+    if attributes.contains_key("db.system") || attributes.contains_key("db.system.name") {
+        let is_cache_op = attributes
```
You could move this "cache op" logic into `derive_db_description`.

loewenheim commented on relay-spans/src/v2_to_v1.rs:
```diff
@@ -270,6 +277,178 @@ fn derive_op_for_v2_span(span: &SpanV2) -> String {
+        return Some(description);
+    }
+
+    let url_path = url_path.unwrap();
```
```suggestion
   let Some(url_path) = url_path else {
        return Some(description);
    };
```

loewenheim commented on relay-spans/src/v2_to_v1.rs:
```diff
@@ -270,6 +277,178 @@ fn derive_op_for_v2_span(span: &SpanV2) -> String {
+        .split('#')
+        .next()
+        .unwrap_or(url)
+        .to_string()
```
`split` lets you pass multiple characters at once:
```suggestion
    url.split(&['?', '#']).next().unwrap_or(url).to_owned()
```

loewenheim commented on relay-spans/src/v2_to_v1.rs:
```diff
@@ -270,6 +277,178 @@ fn derive_op_for_v2_span(span: &SpanV2) -> String {
+        }
+    }
+
+    None
```
In this function there is only the one attribute you're checking, so you don't need the `and_then` and `if let` rigmarole.
```suggestion
    let url = attributes
        .get("url.full")
        .or_else(|| attributes.get("http.url"))?
        .value()?
        .value
        .value
        .value()?
        .as_str()?;

    let parsed_url = Url::parse(url).ok()?;
    Some(format!(
        "{}://{}{}",
        parsed_url.scheme(),
        parsed_url.domain().unwrap_or(""),
        parsed_url.path()
    ))
```

loewenheim commented on relay-spans/src/v2_to_v1.rs:
```diff
@@ -270,6 +277,178 @@ fn derive_op_for_v2_span(span: &SpanV2) -> String {
+        .and_then(|attr_val| attr_val.value.value.value())
+        .and_then(|v| v.as_str())
+    {
+        return Some(query_text.to_string());
```
We generally use `to_owned` for this.
```suggestion
        return Some(query_text.to_owned());
```

gggritso committed:
Simplify condition check

Co-authored-by: Sebastian Zivota <loewenheim@users.noreply.github.com>

gggritso committed:
Simplify conditional

Co-authored-by: Sebastian Zivota <loewenheim@users.noreply.github.com>

gggritso committed:
Prefer `to_owned` over `to_string`

Co-authored-by: Sebastian Zivota <loewenheim@users.noreply.github.com>

gggritso committed:
Simplify split

Co-authored-by: Sebastian Zivota <loewenheim@users.noreply.github.com>

gggritso committed:
Remove redundant comment

gggritso committed:
Simplify conditionals

Allow the inferring function to return `None`, and remove checks from
parent. Also add a test for this.

gggritso committed:
Simplify function

gggritso committed:
Merge branch 'master' into georgegritsouk/ope-37-derive-spandescription-for-otlp-spans

loewenheim reviewed (approved):
There are still some `to_string` left in `v2_to_v1.rs`, otherwise LGTM!

gggritso committed:
Prefer `to_owned` over `to_string`

gggritso committed:
Merge branch 'master' into georgegritsouk/ope-37-derive-spandescription-for-otlp-spans

gggritso committed:
Merge branch 'master' into georgegritsouk/ope-37-derive-spandescription-for-otlp-spans

gggritso committed:
Move changelog to correct spot
